src/Tasks/Microsoft.NET.Build.Tasks/GenerateDepsFile.cs

```diff
@@ -132,7 +134,13 @@ protected override void ExecuteCore()
                 PlatformLibraryName,
                 IsSelfContained);
 
-            DependencyContext dependencyContext = new DependencyContextBuilder(mainProject, projectContext)
+            //  If publishing a framework-dependent app, then include the assembly and file versions of files in the
+            //  deps.json to support rolling forward to a newer shared framework.  See https://github.com/dotnet/core-setup/pull/3704
+            //  Reading these versions could impact build perf, so we avoid doing it if we aren't publishing or if we are
+            //  publishing a self-contained app.
+            bool includeRuntimeFileVersions = IsPublish && !IsSelfContained;
```

Comment:
It may be better to use `projectContext.IsFrameworkDependent` here instead of `!IsSelfContained`.  The comment even says "publishing a framework-dependent app".

Reply:
What do you think about moving this logic out of the C# task all together and making a property for "IncludeRuntimeFileVersions" that is defaulted in the MSBuild XML?  That way someone could turn this on even if they weren't publishing, or in a self-contained app in case they wanted this info for some reason.

Reply:
I think we want to be able to have the info on non-publish optionally.

What Steve and I had considered was:

Default experience:
* Build (F5) -> runtimeconfig.dev.json disables roll forward
* Publish -> writes the info

But there could be an opt-in to have build write the info (taking a perf hit) and not disable roll forward.